File: Python3/1101. The Earliest Moment When Everyone Become Friends.py
class Solution:
    def earliestAcq(self, logs: List[List[int]], n: int) -> int:
        logs.sort(key = lambda x:x[0])
        dsu = DSU(n)
        
        for log in logs:
            dsu.union(log[1], log[2])
            if dsu.getCount() == 1:
                return log[0]
        
        return -1
         
class DSU:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        self.rank = [1]*size
        self.count = size
        
    def find(self, x):
        if x != self.root[x]:
            self.root[x] = self.find(self.root[x])
        return self.root[x]
    
    def union(self, a, b):
        x = self.find(a)
        y = self.find(b)
        
        # different roots, merge the two nodes
        # this means we find another component
        if x != y:
            self.count -= 1
            if self.rank[x] < self.rank[y]:
                self.root[x] = y
            elif self.rank[x] > self.rank[y]:
                self.root[y] = x
            else:
                self.root[x] = y
                self.rank[y] += 1
                
    def getCount(self):
        return self.count

File: Java/1101. The Earliest Moment When Everyone Become Friends.java
class Solution {
    public int earliestAcq(int[][] logs, int n) {
        Arrays.sort(logs, Comparator.comparingInt(o -> o[0]));
        DSU dsu = new DSU(n);
        
        for(int[] log : logs){
            dsu.union(log[1], log[2]);
            if(dsu.getCount() == 1){
                return log[0];
            }
        }
        return -1;
    }
}

class DSU{
    int[] root;
    int[] rank;
    int count;
    
    DSU(int size){
        root = new int[size];
        rank = new int[size];
        count = size;
        
        for(int i=0; i<size; i++){
            root[i] = i;
            rank[i] = 1;
        }
    }
    
    public int find(int x){
        return (x==root[x]) ? x : (root[x]=find(root[x]));
    }
    
    // merge two roots
    public void union(int a, int b){
        int x = find(a);
        int y = find(b);
        
        // different roots, need to merge one to another who has larger rank
        if(x != y){
            count--;
            if(rank[x] < rank[y]){
                root[x] = y;
            }else if(rank[x] > rank[y]){
                root[y] = x;
            }else{
                root[x] = y;
                rank[y]++;
            }
        }
    }
    
    // return the count
    public int getCount(){
        return count;
    }
}